### ISSUE TITLE
[bug] CMakeUserPresets.json gets generated when ConanFile.generators is set

### ISSUE THREAD

**Issue by risa2000** [conan-io/conan#14030]
### Environment details

* Operating System+version: Windows 10 Pro
* Compiler+version: MSVC 193, VS2022
* Conan version: 1.60.1
* Python version:  3.10.10

### Steps to reproduce

using this as `conanfile.py`:
```python
import os.path

from conan import ConanFile
from conan.tools.layout import cmake_layout
from conan.tools.cmake import CMake, CMakeToolchain, CMakeDeps
from conan.tools.files import collect_libs, rmdir

required_conan_version = ">=1.51.0"

class TestConan(ConanFile):
    name = "test"
    description = "Testing"
    requires = [
        "fmt/9.1.0",
        "spdlog/1.11.0",
        "nlohmann_json/3.11.2",
    ]

    settings = "os", "arch", "compiler", "build_type"

    generators = "CMakeToolchain"

    def build(self):
        cmake = CMake(self)
        cmake.configure()
        cmake.build()

    def generate(self):
        tc = CMakeToolchain(self)
        tc.user_presets_path = False
        tc.generate()

        deps = CMakeDeps(self)
        deps.generate()

    def package(self):
        cmake = CMake(self)
        cmake.install()
        rmdir(self, os.path.join(self.package_folder, "lib", "cmake")) # remove native CMake package files

    def package_info(self):
        self.cpp_info.names["generator_name"] = "test"
        self.cpp_info.includedirs = ['include']  # Ordered list of include paths
        self.cpp_info.libs = collect_libs(self)  # The libs to link against
        self.cpp_info.system_libs = []  # System libs to link against
        self.cpp_info.libdirs = ['lib']  # Directories where libraries can be found
        self.cpp_info.resdirs = ['res']  # Directories where resources, data, etc. can be found
        self.cpp_info.bindirs = ['bin']  # Directories where executables and shared libs can be found
        self.cpp_info.srcdirs = []  # Directories where sources can be found (debugging, reusing sources)
        self.cpp_info.build_modules = {}  # Build system utility module files
        self.cpp_info.defines = []  # preprocessor definitions
        self.cpp_info.cflags = []  # pure C flags
        self.cpp_info.cxxflags = []  # C++ compilation flags
        self.cpp_info.sharedlinkflags = []  # linker flags
        self.cpp_info.exelinkflags = []  # linker flags
```
with empty `CMakeLists.txt` and run as:
```cmd
conan install -if out/conan -s build_type=Debug -s compiler.cppstd=20 .
```
produces `CMakeUserPresets.json`, even though it is explicitly disabled in `generate` function.

`CMakeUserPresets.json` is not produced when, either:
* install folder is not specified (i.e. the files are generated into the same folder as `conanfile.py` and `CMakeLists.txt` are stored)
* I remove the line `generators = "CMakeToolchain` from `conanfile.py`.

I am not sure if it is a bug or not, but seems inconsistent at least, plus `CMakeUserPresets.json` is really annoying as it breaks Visual Studio IDE.

**Comment by memsharded** [comment]
Hi @risa2000 

Having both ``generators = "CMakeToolchain"`` and then declaring ``toolchain = CMakeToolchain(self)`` in the ``generate()`` method is not possible, you should be using one or the other. Conan 2.0 will give you an error message and stop in this case, but 1.60 couldn't implement this check yet.

So if you remove ``generators = "CMakeToolchain"`` everything works, then this would be the expected behavior.

**Comment by risa2000** [comment]
@memsharded
Actually, it is perfectly possible to have both, and it worked in all other aspects (apart from the `CMakeUserPresets.json` generation control).

The doc says that it is possible to use one or the other, but does not say what happens when both are used (or whether one will overshadow the other, etc.).
https://docs.conan.io/1/reference/conanfile/tools/cmake/cmaketoolchain.html
It also mentions that the file is created when `layout` is used, which is not the case here either, but I guess it might be another artefact of the use of `generators` variable?

Anyway, I understand that this is not a supported use. It is a pity that this is not caught at least in runtime as I was pulling my hair over it for some time and effectively found out only by accident.
It is good to hear that 2.0 handles it differently, unfortunately, 2.0 is not a viable solution for me at the moment.

On the side note. The way the generation is controlled is unfortunate as it implies that changing the behavior requires changing the `conanfile.py` file. I believe this should be parametrized as the fact/requirement to generate or avoid generation of `CMakeUserPresets.json` is not property of the package but comes from the environment. E.g. in Visual Studio IDE I might not want it, but on a different platform or in a different IDE I could have good use of it.

**Comment by memsharded** [comment]
> The doc says that it is possible to use one or the other, but does not say what happens when both are used (or whether one will overshadow the other, etc.).

Yes, that is typical, the docs document the possibilities, but do not always document all possible forbidden behaviors. In this case, declaring it in both places is not allowed, as said, it will raise a hard error in 2.0.

> On the side note. The way the generation is controlled is unfortunate as it implies that changing the behavior requires changing the conanfile.py file. I believe this should be parametrized as the fact/requirement to generate or avoid generation of CMakeUserPresets.json is not property of the package but comes from the environment.

Yes, I agree that this might benefit from a ``conf`` that control the behavior. In practice, it is usually a pragmatic approach just to let Conan generate the file it and ``gitignore`` it.

**Comment by risa2000** [comment]
@memsharded 
I believe there is some misunderstanding about what exactly generated `CMakeUserPresets.json` does in Visual Studio IDE.
which I will try to clarify:

VS IDE has been historically using `CMakeSettings.json` (before `CMakePresets.json` was introduced), which is very similar in its purpose to `CMakePresets.json`, but not the same. Recent versions of VS IDE can use `CMakePresets.json` (or `CMakeUserPresets.json`) as well, and it is now recommended approach, but `CMakeSettings.json` is still supported. If VS founds both to be present it prefers `CMakePresets.json`.

When I open a CMake project in VS, which uses `CMakeSettings.json` and then run `conan install`, which generates `CMakeUserPresets.json` the next user action in IDE will break the IDE configuration to the point it can no longer select the build and the only way to recover (after deleting `CMakeUserPresets.json` ) is by deleting the IDE cache folder `.vs`, which can only be done after quitting the IDE. So this is quite a hassle.
Setting `.gitignore` has no impact on this behavior as it is not a `git` problem, but IDE problem.

**Comment by wayneph01** [comment]
@risa2000 I suspect the issue you're having with Presets breaking the IDE is the same thing I noticed trying to extend presets in one of the Examples.  (#13968)

In your CMakePresets.json file, I suspect you have multiple presets referencing the same "conan-buildtype" reference so the CMakeUserPresets.json file is generating duplicates causing the IDE and CMake not to be able to use it.

**Comment by risa2000** [comment]
@wayneph01 I am using `CMakeSettings.json`, which is similar but not compatible to `CMakePresets.json`, and I believe the IDE cannot handle starting with one and then being handed the other during the build.

**Comment by memsharded** [comment]
Conan provided some time ago the ``tools.cmake.cmaketoolchain:user_presets`` that can be assigned to "" empty string, and that will completely disable the generation of ``CMakeUserPresets.json``. This can also be done from the ``CMakeToolchain.user_presets_path=False`` attribute in the recipe ``generate()`` method.

I am closing this ticket as solved, please create new ticket for any further question, issue or comment that you might have. Thanks for your feedback!
